File: apps/python/lib/chapter15/training.py
from pathlib import Path

from lib.chapter02.iris_preprocessing import split_train_test
from lib.chapter07.cinema_regression import fit_linear_regression, prepare_cinema
from lib.chapter08.survived_classifier import (
    build_pipeline,
    load_survived,
    split_features_and_target,
)
from lib.chapter15.infrastructure import JoblibModelStore

TEST_SIZE = 0.2
SEED = 0
MAX_DEPTH = 5


def train_and_save_models(data_directory: Path, store: JoblibModelStore) -> None:
    cinema = prepare_cinema(
        data_directory / "cinema.csv", test_size=TEST_SIZE, seed=SEED
    )
    store.save_sales_model(fit_linear_regression(cinema.x_train, cinema.t_train))

    x, t = split_features_and_target(load_survived(data_directory / "Survived.csv"))
    survived = split_train_test(x, t, test_size=TEST_SIZE, seed=SEED)
    pipeline = build_pipeline(max_depth=MAX_DEPTH, class_weight="balanced")
    store.save_survival_model(pipeline.fit(survived.x_train, survived.t_train))

File: apps/kotlin/src/main/kotlin/chapter15/Training.kt
package chapter15

import chapter02.splitTrainTest
import chapter07.fitLinearRegression
import chapter07.prepareCinema
import chapter08.ClassWeight
import chapter08.buildPipeline
import chapter08.loadSurvived
import chapter08.splitFeaturesAndTarget
import java.io.File

private const val TEST_SIZE = 0.2
private const val SEED = 0
private const val MAX_DEPTH = 5

/** 第 7・8 章と同じ条件でモデルを学習し、置き場に保存する */
fun trainAndSaveModels(
    dataDirectory: File,
    store: FileModelStore,
) {
    val cinema = prepareCinema(File(dataDirectory, "cinema.csv"), testSize = TEST_SIZE, seed = SEED)
    store.saveSalesModel(fitLinearRegression(cinema.xTrain, cinema.tTrain))

    val (x, t) = splitFeaturesAndTarget(loadSurvived(File(dataDirectory, "Survived.csv")))
    val survived = splitTrainTest(x, t, testSize = TEST_SIZE, seed = SEED)
    val pipeline = buildPipeline(maxDepth = MAX_DEPTH, classWeight = ClassWeight.BALANCED)
    store.saveSurvivalModel(pipeline.fit(survived.xTrain, survived.tTrain))
}
